### PR TITLE
Add option to restrict the maximum hub message size

### PR DESCRIPTION
**Pull request by davidfowl**
- This change moves the limit checking from the transport layer to the protocol parsing layer. One nice side effect is that it gives us better control over error handling.

Fixes https://github.com/aspnet/AspNetCore/issues/8121 and https://github.com/aspnet/AspNetCore/issues/5307

Notes:
- The transport limits just slow down the client, but it doesn't prevent extra buffering. See the issue for more details.

### CODE REVIEW

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTestUtils/Utils.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ public static IServiceProvider CreateServiceProvider(Action<ServiceCollection> a
             return services.BuildServiceProvider();
         }
 
-        public static Connections.ConnectionHandler GetHubConnectionHandler(Type hubType, Action<ServiceCollection> addServices = null, ILoggerFactory loggerFactory = null)
+        public static Connections.ConnectionHandler GetHubConnectionHandler(Type hubType, ILoggerFactory loggerFactory = null, Action<ServiceCollection> addServices = null)
```
Comment (davidfowl) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTestUtils/Utils.cs]: Lambdas should be last.

File: src/SignalR/server/Core/src/HubOptions.cs [hunk 2]
```diff
@@ -36,6 +36,11 @@ public class HubOptions
         /// </summary>
         public IList<string> SupportedProtocols { get; set; } = null;
 
+        /// <summary>
+        /// Gets or sets the maximum message size of a single hub message.
+        /// </summary>
+        public long? MaximumMessageSize { get; set; } = null;
```
Comment (JamesNK) [inline src/SignalR/server/Core/src/HubOptions.cs]: This is just for receive, right? Should the name be MaximumReceiveMessageSize?

File: src/SignalR/server/Core/src/Internal/HubOptionsSetup.cs [hunk 3]
```diff
@@ -16,6 +16,8 @@ public class HubOptionsSetup : IConfigureOptions<HubOptions>
 
         internal static TimeSpan DefaultClientTimeoutInterval => TimeSpan.FromSeconds(30);
 
+        internal const int DefaultMaximumMessageSize = 32 * 1024 * 1024;
```
Comment (JamesNK) [inline src/SignalR/server/Core/src/Internal/HubOptionsSetup.cs]: Include the default size in the HubOptions.MaximumMessageSize XML doc?

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 4]
```diff
@@ -452,6 +453,191 @@ public async Task HandshakeSuccessSendsResponseWithoutError()
...
+                    await connectionHandlerTask.OrTimeout();
+                }
+            }
+
+            Assert.NotNull(exception);
+            Assert.Equal(exception.Message, $"The maximum message size of {maximumMessageSize}B was exceeded. The message size can be configured in AddHubOptions.");
+        }
+
+        [Fact]
+        public async Task ManyHubMessagesUnderTheMessageSizeButConfiguredWithMax()
```
Comment (JamesNK) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: good test 👍

File: src/SignalR/server/Core/src/Internal/HubOptionsSetup.cs (hunk 3 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/HubOptionsSetup.cs]: Done

**Review by JamesNK — APPROVED** [review]

File: src/SignalR/server/Core/src/HubOptions.cs [hunk 5]
```diff
@@ -36,6 +36,11 @@ public class HubOptions
         /// </summary>
         public IList<string> SupportedProtocols { get; set; } = null;
 
+        /// <summary>
+        /// Gets or sets the maximum message size of a single incoming hub message. The default is 32KB.
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/HubOptions.cs]: "Null means no maximum message size" or is that implicit with nullable types

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**